vault/provider_test.go

```diff
@@ -618,54 +618,131 @@ func TestAccTokenName(t *testing.T) {
-					Check:  testTokenName_check(test.WantTokenName),
+					Config: testProviderConfig(test.UseTokenNameSchema, `token_name = "`+test.TokenNameSchema+`"`),
+					Check:  checkTestSecretData("display_name", test.WantTokenName),
 				},
 			},
+			CheckDestroy: testEnvCleanup("VAULT_TOKEN_NAME"),
+		})
+	}
+}
+
+func TestAccChildToken(t *testing.T) {
+
+	tests := []struct {
+		ChildTokenEnv        string
+		TestChildTokenEnv    bool
+		ChildTokenSchema     string
+		TestChildTokenSchema bool
+		ExpectChildToken     bool
+	}{
+		{
+			TestChildTokenSchema: false,
+			TestChildTokenEnv:    false,
+			ExpectChildToken:     true,
+		},
+		{
+			ChildTokenEnv:     "",
+			TestChildTokenEnv: true,
+			ExpectChildToken:  true,
+		},
+		{
+			ChildTokenEnv:     "true",
+			TestChildTokenEnv: true,
+			ExpectChildToken:  true,
+		},
+		{
+			ChildTokenEnv:     "false",
+			TestChildTokenEnv: true,
+			ExpectChildToken:  false,
+		},
+		{
+			ChildTokenSchema:     "true",
+			TestChildTokenSchema: true,
+			ExpectChildToken:     true,
+		},
+		{
+			ChildTokenSchema:     "false",
+			TestChildTokenSchema: true,
+			ExpectChildToken:     false,
+		},
+		{
+			ChildTokenEnv:        "true",
+			TestChildTokenEnv:    true,
+			ChildTokenSchema:     "false",
+			TestChildTokenSchema: true,
+			ExpectChildToken:     false,
+		},
+		{
+			ChildTokenEnv:        "false",
+			TestChildTokenEnv:    true,
+			ChildTokenSchema:     "true",
+			TestChildTokenSchema: true,
+			ExpectChildToken:     true,
+		},
+	}
+
+	for _, test := range tests {
+		resource.Test(t, resource.TestCase{
+			Providers: testProviders,
+			PreCheck:  func() { testAccPreCheck(t) },
+			Steps: []resource.TestStep{
+				{
+					PreConfig: func() {
+						if test.TestChildTokenEnv {
+							err := os.Setenv("TERRAFORM_VAULT_CREATE_CHILD_TOKEN", test.ChildTokenEnv)
+							if err != nil {
+								t.Fatal(err)
+							}
+						} else {
+							err := os.Unsetenv("TERRAFORM_VAULT_CREATE_CHILD_TOKEN")
+							if err != nil {
+								t.Fatal(err)
+							}
+						}
+					},
+					Config: testProviderConfig(test.TestChildTokenSchema, `create_intermediate_child_token = `+test.ChildTokenSchema),
+					Check:  checkUsedToken(test.ExpectChildToken),
+				},
+			},
+			CheckDestroy: testEnvCleanup("TERRAFORM_VAULT_CREATE_CHILD_TOKEN"),
 		})
 	}
 }
 
 func testHeaderConfig(headerName, headerValue string) string {
 	providerConfig := fmt.Sprintf(`
-	provider "vault" {
 		headers {
 			name  = "%s"
 			value = "%s"
 		}
 		token_name = "testtoken"
-	}
-
-	data "vault_generic_secret" "test" {
-		path = "/auth/token/lookup-self"
-	}
 	`, headerName, headerValue)
-	return providerConfig
+	return testProviderConfig(true, providerConfig)
 }
 
 // Using the data lookup generic_secret to inspect used token
 // by terraform (this enables check of token name)
-func testTokenNameConfig(tokenNameSchema bool, tokenName string) string {
+func testProviderConfig(includeProviderConfig bool, config string) string {
```

Comment:
> Please update the documentation to match.

Can you be more clear about what documentation needs updating here? The comment above this function is still 100% accurate, and the change to make this more generic didn't affect anything else.

Reply:
as for your second comment:

There are several tests that call this function. One of those tests always includes a header provider config, the others will only want the provider configuration if the test is specifically testing the provider schema.

To be fair, I think I agree that simply passing in an empty provider config string would be more logical, and possibly nicer within this function, but it would require moving that specific logic (to determine if the provider config should be an empty string or actually contain the schema under test) elsewhere, and then we're back to this same conditional. If you would prefer I separate this into two functions, one for generating the Terraform config string and one for evaluating a conditional to decide if the provider portion of that Terraform config should including something or not, let me know, but I've left this as-is for now, since it doesn't really make sense to me.

In case the confusion you had was around the naming of this function (eg. why does `testProviderConfig` sometimes not include the "Provider" portion of the config and only a data resource?) I changed the name to `getTerraformConfig` which reduces that confusion at the expense of being a bit generic.

Reply:
The extra boolean is argument is redundant with the config string.

`if config != "" {...}`

Reply:
Take a look at line 621: https://github.com/hashicorp/terraform-provider-vault/pull/775/files#diff-96e3c1e3f2b69ac273388431a9cebb277d7a5d596691beb0a32b5821c5d6b063R621
and 703: https://github.com/hashicorp/terraform-provider-vault/pull/775/files#diff-96e3c1e3f2b69ac273388431a9cebb277d7a5d596691beb0a32b5821c5d6b063R703

In both of those calls, config is never `""`. So this is NOT redundant.

I can refactor this (and the existing tests that use it) even further if you only want this function to have one argument, but to keep it clean and readable, I will probably need to add 2 additional functions then call this `getTerraformConfig` in order to keep each test's logic isolated. Is that what you are asking for?

Reply:
Thanks, but I think it is okay to keep things as they are, but `getTerraformConfig` should be renamed as suggested earlier.

Reply:
Thanks, but I think it is okay to keep things as they are, but `getTerraformConfig` should be renamed as suggested earlier.